lib/functions/main/build-tasks.sh

```diff
@@ -0,0 +1,281 @@
+				FULL=""
+				REPLACE="-full"
+				compile_firmware
+				FULL="-full"
+				REPLACE=""
+				compile_firmware
+			}
+
+		fi
+
+	fi
+}
+
+build_armbian-bsp() {
+	# create board support package
+	[[ -n "${RELEASE}" && ! -f "${DEB_STORAGE}/${BSP_CLI_PACKAGE_FULLNAME}.deb" && "${REPOSITORY_INSTALL}" != *armbian-bsp-cli* ]] && create_board_package
+
+	# create desktop package
+	[[ -n "${RELEASE}" && "${DESKTOP_ENVIRONMENT}" && ! -f "${DEB_STORAGE}/$RELEASE/${CHOSEN_DESKTOP}_${REVISION}_all.deb" && "${REPOSITORY_INSTALL}" != *armbian-desktop* ]] && create_desktop_package
+	[[ -n "${RELEASE}" && "${DESKTOP_ENVIRONMENT}" && ! -f "${DEB_STORAGE}/${RELEASE}/${BSP_DESKTOP_PACKAGE_FULLNAME}.deb" && "${REPOSITORY_INSTALL}" != *armbian-bsp-desktop* ]] && create_bsp_desktop_package
+}
+
+build_chroot() {
+	# build additional packages
+	[[ $EXTERNAL_NEW == compile ]] && chroot_build_packages
+}
+
+build_bootstrap() {
+	# These two keys are necessary for backward compatibility with logic
+	# https://github.com/armbian/scripts/tree/master/.github/workflows scripts.
+	# They need to be removed when the need disappears there.
+	if [[ $KERNEL_ONLY != yes ]]; then
+		[[ $BSP_BUILD != yes ]] && debootstrap_ng
+	fi
+}
+
+#################################################################################################################################
+#
+# build_main()
+#
+# Builds all artifacts or the filtered ones only.
+# Ensures that any build pre-requisite is met.
+#
+# $1: optional comma separated list of artifacts to build only.
+#     If this list is empty or not set, then all build tasks will be performed.
+#     The following build task names are supported for filtering build tasks:
+#       u-boot, kernel, armbian-config, armbian-zsh, plymouth-theme-armbian, armbian-firmware, armbian-bsp, chroot, bootstrap
+#
+build_main() {
+	local _buildOnly=$1
+	local _kernel_buildOnly="u-boot kernel armbian-config armbian-zsh plymouth-theme-armbian armbian-firmware armbian-bsp"
+	local _all_valid_buildOnly="u-boot kernel armbian-config armbian-zsh plymouth-theme-armbian armbian-firmware armbian-bsp chroot bootstrap"
+
+	start=$(date +%s)
+
+	# These checks are necessary for backward compatibility with logic
+	# https://github.com/armbian/scripts/tree/master /.github/workflows scripts.
+	# They need to be removed when the need disappears there.
+	[[ -n $KERNEL_ONLY ]] && {
+		display_alert "The KERNEL_ONLY key is no longer used." "KERNEL_ONLY=$KERNEL_ONLY" "wrn"
+		if [ "$KERNEL_ONLY" == "no" ]; then
+			display_alert "use an empty BUILD_ONLY variable" "" "info"
+			[[ -n "${_buildOnly}" ]] && {
+				display_alert "A contradiction. BUILD_ONLY contains a goal. Fix it." "${_buildOnly}" "wrn"
+				_buildOnly=""
+			}
+		elif [ "$KERNEL_ONLY" == "yes" ]; then
+			display_alert "Instead, use BUILD_ONLY to select the build target." "$_kernel_buildOnly" "wrn"
+			_buildOnly="$_kernel_buildOnly"
+		fi
```

Comment:
Here is the problem. We set `_buildOnly` but didn't set `BUILD_ONLY`. The check in other function will error

Reply:
Sure - understood - will refactor this accordingly and let you knw.

Reply:
No. BUILD_ONLY is a global variable. It is initialized at the very beginning.

Reply:
We need to think about how to make it beautiful. 
@hzyitc thanks.